File: dsa-learning/dsa-java-code/src/main/java/com/dsa/leetcode/labuladong/binary_search/P34_FindFirstAndLastPositionOfElementInSortedArray.java
package com.dsa.leetcode.labuladong.binary_search;

/**
 * @since 2023/11/21 22:19
 */
public class P34_FindFirstAndLastPositionOfElementInSortedArray {

    public int[] searchRange(int[] nums, int target) {
        if (nums == null || nums.length == 0) {
            return new int[]{-1, -1};
        }
        int first = firstBinarySearch(nums, target);
        int last = lastBinarySearch(nums, target);
        return new int[]{first, last};
    }

    private int lastBinarySearch(int[] nums, int target) {
        int left = 0, right = nums.length - 1;
        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (nums[mid] == target) {
                left = mid + 1;
            } else if (nums[mid] > target) {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        if (right < 0) {
            return -1;
        }
        return nums[right] == target ? right : -1;
    }

    private int firstBinarySearch(int[] nums, int target) {
        int left = 0, right = nums.length - 1;
        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (nums[mid] == target) {
                right = mid - 1;
            } else if (nums[mid] > target) {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        if (left >= nums.length) {
            return -1;
        }
        return nums[left] == target ? left : -1;
    }
}

File: dsa-learning/dsa-python-code/com/dsa/leetcode/labuladong/binary_search/P34_FindFirstAndLastPositionOfElementInSortedArray.py
# author: tobebetter9527
# since: 2023/11/22 9:04
from typing import List


class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        if not nums:
            return [-1, -1]
        first = self.firstBinarySearch(nums, target)
        last = self.lastBinarySearch(nums, target)
        return [first, last]

    def firstBinarySearch(self, nums, target):
        left, right = 0, len(nums) - 1
        while left <= right:
            mid = left + (right - left) // 2
            if nums[mid] == target:
                right = mid - 1
            elif nums[mid] > target:
                right = mid - 1
            else:
                left = mid + 1
        if left >= len(nums):
            return -1
        return left if nums[left] == target else -1

    def lastBinarySearch(self, nums, target):
        left, right = 0, len(nums) - 1
        while left <= right:
            mid = left + (right - left) // 2
            if nums[mid] == target:
                left = mid + 1
            elif nums[mid] > target:
                right = mid - 1
            else:
                left = mid + 1
        if right < 0:
            return -1
        return right if nums[right] == target else -1


if __name__ == '__main__':
    nums = [5, 7, 7, 8, 8, 10]
    target = 8
    solu = Solution()
    print(solu.searchRange(nums, target))
